### ISSUE TITLE
Enabler: Build Drupal NCIDS Banner + Language Toggle

### ISSUE THREAD

**Issue by bryanpizzillo** [NCIOCPL/cgov-digital-platform#3238]
# Build NCIDS Banner + Language Toggle

Remaining work for this ticket is implementing analytics.

> **ESTIMATE** 2

This is a placeholder ticket for the building and implementation of the NCIDS Banner + Language Toggle as an element in NCIDS.  

### Language Toggle
The Drupal language toggle has a lot of baggage that comes with it.
1. It prevents the page from being cached in the [Dynamic Page Cache](https://www.drupal.org/project/drupal/issues/2232375) This has other issues affecting the purge module as well, since the newest purge modules use the DPC to determine the page's cachability.
2. The language toggle is a block, which means it cannot appear in a raw html block. This limits our flexibility with drawing the banner.
3. OOB the language toggle is actually an unordered list even if there is only one link
4. OOB the language toggle actually shows the toggle no matter if the page is translated or not. We made our own hacks to get around this. 

As such, I would like to avoid using the Drupal language toggle altogether if possible. Instead we can leverage the hreflang links and have the toggle be Javascript-based. 

#### Approach:
1. Create an object (let's call it LanguageLabel) with the 2 character language code, and a label of the text for a link. e.g.
    ```
    {
      en: 'English',
      es: 'Español',
    ```
1. on dom loaded or something, Look for the `<link rel="alternate" hreflang="<lang>" href="<URL>" />` elements in the head block.
2. Filter out links with the same lang as the `<html>` element.
3. Filter out links that do not have a key in LanguageLabel
4. If there are still any links left, generate the link using the following information:
    1. Href: use the link's href element
    2. Text: get the value of LanguageLabel[hreflang of the link] 

### Banner Bar
There are a few possible approaches for implementing:
1. Create a new region to hold the banner (this will always be needed)
2. then either:
    1. Create a block and hardcode the text into the template for the block.
        * Updating the contents of the block across all sites is automatic.
        * Block can be added to new sites automatically through the theme configuration. (Kids play stuff)
        * Translation will be a pain.
        * Changes to block require a release.
    2. create it as a custom library block (aka raw html). 
        * under emergency situations or something, contents can be updated without a release
        * translations work through the UI and are as easy as pasting the new text
        * Updates to the block must be done for each site
        * Block can be automatically added, but with effort. 
           * Changes to text would still require a release to make sure new sites do not get old content.

_If the assumption is that this block will never really change and all sites have the same contents, then option 1 is the best approach._

### Acceptance Criteria for Behaviors 
*If a component changes between breakpoints, or things show/hide, that gets captured here* 

```gherkin
Scenario: User is on an English page of c.gov on desktop
  Given user is on a page with a translation equivalent
  Then the language toggle "Español" button will appear in the top right on the usa banner component
  When the user hovers over the button
  Then the hover state is applied to the button
  When the user clicks on the button
  Then the button links to the translation of that page 
```

```gherkin
Scenario: User is on a Spanish page of c.gov on desktop
  Given user is on a page with a translation equivalent
  Then the language toggle "English" link will appear in the top right on the usa banner component
  When the user hovers over the button
  Then the hover state is applied to the button
  When the user clicks on the button
  Then the button links to the translation of that page 
```

```gherkin
Scenario: User is on an English page of c.gov on mobile
  Given user is on a page with a translation equivalent
  Then the language toggle "Español" link will appear in the top right on the usa banner component with a pipe to the left separating the official website text
    And the text will link to the translation of that page
```

```gherkin
Scenario: User is on a Spanish page of c.gov on mobile
  Given user is on a page with a translation equivalent
  Then the language toggle "English" link will appear in the top right on the usa banner component with a pipe to the left separating the official website text
    And the text will link to the translation of that page
```

```gherkin
Scenario: User is on a page of c.gov (all breakpoints) without a translation equivalent
  Given user is on a page without a translation equivalent
  Then no language toggle link will appear in the top right on the usa banner component
```

### Acceptance Criteria for Analytics
``` gherkin
Scenario: User on English page clicks the language toggle to view Spanish page
Given a user is on an English page of cancer.gov that has a Spanish translation
When the user clicks on the "español" link within the banner at the top of the page
Then there should be an NCIDataLayer Other event 
  And the event has the name “PreHeader:LinkClick”
  And the event has the the linkname “PreHeader:LinkClick”
  And this event should have the following data:
      | key              | value              |
      | linkText         | English to Spanish |
      | location         | PreHeader          |
      | action           | Link Click         |
      | preHeaderElement | Language           |
```

``` gherkin
Scenario: User on Spanish page clicks the language toggle to view English page
Given  a user is on a Spanish page of cancer.gov that has an English translation
When the user clicks on the "English" link within the banner at the top of the page
Then there should be an NCIDataLayer Other event 
  And the event has the name “PreHeader:LinkClick”
  And the event has the the linkname “PreHeader:LinkClick”
  And this event should have the following data:
      | key              | value              |
      | linkText         | Spanish to English |
      | location         | PreHeader          |
      | action           | Link Click         |
      | preHeaderElement | Language           |
```
--------

### Prerequisites
- [ ] NCIOCPL/ncids#261

**Comment by bryanpizzillo** [comment]
That was a bunch of nice words above. So we are just going to copy the LanguageBlock and be done with it. This will reduce the JS but also not add in any CLS issues or other visual quirks. Also it will get done faster. :)

**Comment by meglake** [comment]
@sarinapadilla will add analytics to this ticket and then this ticket will be put into the backlog for storypointing.

**Comment by bryanpizzillo** [comment]
@welshja Here is the ticket for the Banner + Language Toggle. IDK If you still want to call the event "Banner:LinkClick" since you said you want this event for Site Alerts as well, which at this time, are not exactly slated for development, but I guess have to be. (The USWDS component is available, we are just not making it in Drupal)

**Comment by welshja** [comment]
@bryanpizzillo - updated values for linktext on the english and spanish buttons.

**Comment by sarinapadilla** [comment]
Closing this ticket as it has been fixed by PR #3393!
